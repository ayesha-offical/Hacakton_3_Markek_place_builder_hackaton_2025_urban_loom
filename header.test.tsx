import React from "react";
import { render, screen } from "@testing-library/react";
import Header from "./components/header"; // Adjust path if necessary
import { currentUser } from "@clerk/nextjs/server";
import "@testing-library/jest-dom";

// --- Mocks ---

// Mock next/image to simply render an <img>
jest.mock("next/image", () => (props: any) => {
  // eslint-disable-next-line jsx-a11y/alt-text
  return <img {...props} />;
});

// Mock next/form to render a plain <form>
jest.mock("next/form", () => ({
  __esModule: true,
  default: ({ children }: { children: React.ReactNode }) => <form>{children}</form>,
}));

// Mock Container to render its children
jest.mock("./components/container", () => ({
  __esModule: true,
  default: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
}));

// Mock Cart component as a simple stub with a test id
jest.mock("./components/cart", () => () => <div data-testid="cart">Cart</div>);

// Mock Clerk components from @clerk/nextjs
jest.mock("@clerk/nextjs", () => ({
  ClerkLoaded: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  SignedIn: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  SignInButton: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  UserButton: () => <div>UserButton</div>,
}));

// Mock currentUser from @clerk/nextjs/server
jest.mock("@clerk/nextjs/server", () => ({
  currentUser: jest.fn(),
}));

// --- Tests ---

describe("Header Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renders header with signed in user", async () => {
    // Simulate a signed in user
    (currentUser as jest.Mock).mockResolvedValue({ fullName: "John Doe" });
    
    // Await the async Header component
    const HeaderComponent = await Header();
    render(HeaderComponent);
    
    // Wait for the user name to appear
    expect(await screen.findByText("John Doe")).toBeInTheDocument();
    expect(screen.getByTestId("cart")).toBeInTheDocument();
    // Verify that the orders link is present
    expect(screen.getByRole("link", { name: /orders/i })).toBeInTheDocument();
    
    // Snapshot test
    expect(document.body).toMatchSnapshot();
  });

  test("renders header with no user (shows SignInButton)", async () => {
    // Simulate no signed in user
    (currentUser as jest.Mock).mockResolvedValue(null);
    
    // Await the async Header component
    const HeaderComponent = await Header();
    render(HeaderComponent);
    
    // Wait for the SignInButton branch to show text "Login"
    expect(await screen.findByText("Login")).toBeInTheDocument();
    expect(screen.getByTestId("cart")).toBeInTheDocument();
    
    // Snapshot test
    expect(document.body).toMatchSnapshot();
  });
});
