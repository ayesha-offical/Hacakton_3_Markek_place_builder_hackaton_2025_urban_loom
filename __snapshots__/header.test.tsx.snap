// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component renders header with no user (shows SignInButton) 1`] = `
<body>
  <div>
    <header
      class="w-full py-4 border-b border-gray-400 sticky top-0 z-50 bg-white "
    >
      <div>
        <a
          class="font-bold text-3xl hover:text-gray-600"
          href="/"
        >
          <img
            alt="Urban Loom Logo"
            height="60"
            src="/Urban Loom.png"
            width="200"
          />
        </a>
        <form>
          <input
            class="w-full Font- font-light text-black  border border-gray-400 rounded-full px-4 py-2.5 "
            name="query"
            placeholder="Search for a store "
            type="text"
          />
        </form>
        <div
          class="flex items-center gap-5"
        >
          <div
            data-testid="cart"
          >
            Cart
          </div>
          <div>
            <div>
              <a
                class="flex items-center text-sm gap-2 border border-gray-200 px-2 py-1 rounded-md shadow-md hover:shadow-none hoverEffect"
                href="/orders"
              >
                <svg
                  class="lucide lucide-shopping-basket text-cyan-950 w-5 h-5 md:w-6 md:h-6 hoverEffect"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m15 11-1 9"
                  />
                  <path
                    d="m19 11-4-7"
                  />
                  <path
                    d="M2 11h20"
                  />
                  <path
                    d="m3.5 11 1.6 7.4a2 2 0 0 0 2 1.6h9.8a2 2 0 0 0 2-1.6l1.7-7.4"
                  />
                  <path
                    d="M4.5 15.5h15"
                  />
                  <path
                    d="m5 11 4-7"
                  />
                  <path
                    d="m9 11 1 9"
                  />
                </svg>
                <div
                  class="flex flex-col "
                >
                  <p
                    class="text-xs"
                  >
                    <span
                      class="font-semibold"
                    >
                      0 
                    </span>
                    items
                  </p>
                  <p
                    class="font-semibold md:text-base text-xs"
                  >
                    Orders
                  </p>
                </div>
              </a>
            </div>
            <div>
              <div
                class="flex items-center text-sm gap-2 border border-gray-200 px-2 py-1 rounded-md shadow-md hover:shadow-none hoverEffect"
              >
                <svg
                  class="lucide lucide-user"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                  />
                  <circle
                    cx="12"
                    cy="7"
                    r="4"
                  />
                </svg>
                <div
                  class="flex flex-col "
                >
                  <p
                    class="text-xs"
                  >
                    Account
                  </p>
                  <p
                    class="font-semibold"
                  >
                    Login
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
  </div>
</body>
`;

exports[`Header Component renders header with signed in user 1`] = `
<body>
  <div>
    <header
      class="w-full py-4 border-b border-gray-400 sticky top-0 z-50 bg-white "
    >
      <div>
        <a
          class="font-bold text-3xl hover:text-gray-600"
          href="/"
        >
          <img
            alt="Urban Loom Logo"
            height="60"
            src="/Urban Loom.png"
            width="200"
          />
        </a>
        <form>
          <input
            class="w-full Font- font-light text-black  border border-gray-400 rounded-full px-4 py-2.5 "
            name="query"
            placeholder="Search for a store "
            type="text"
          />
        </form>
        <div
          class="flex items-center gap-5"
        >
          <div
            data-testid="cart"
          >
            Cart
          </div>
          <div>
            <div>
              <a
                class="flex items-center text-sm gap-2 border border-gray-200 px-2 py-1 rounded-md shadow-md hover:shadow-none hoverEffect"
                href="/orders"
              >
                <svg
                  class="lucide lucide-shopping-basket text-cyan-950 w-5 h-5 md:w-6 md:h-6 hoverEffect"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m15 11-1 9"
                  />
                  <path
                    d="m19 11-4-7"
                  />
                  <path
                    d="M2 11h20"
                  />
                  <path
                    d="m3.5 11 1.6 7.4a2 2 0 0 0 2 1.6h9.8a2 2 0 0 0 2-1.6l1.7-7.4"
                  />
                  <path
                    d="M4.5 15.5h15"
                  />
                  <path
                    d="m5 11 4-7"
                  />
                  <path
                    d="m9 11 1 9"
                  />
                </svg>
                <div
                  class="flex flex-col "
                >
                  <p
                    class="text-xs"
                  >
                    <span
                      class="font-semibold"
                    >
                      0 
                    </span>
                    items
                  </p>
                  <p
                    class="font-semibold md:text-base text-xs"
                  >
                    Orders
                  </p>
                </div>
              </a>
            </div>
            <div
              class="flex items-center text-sm gap-2 border border-gray-200 px-2 py-1 rounded-md shadow-md hover:shadow-none hoverEffect"
            >
              <div>
                UserButton
              </div>
              <div
                class="hidden md:inline-flex flex-col "
              >
                <p
                  class="text-xs"
                >
                  Account Login!
                </p>
                <p
                  class="font-semibold"
                >
                  John Doe
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
  </div>
</body>
`;
